lib/hash_ring/ets.ex

```diff
@@ -41,202 +59,220 @@ defmodule ExHashRing.HashRing.ETS do
   end
 
-  @spec remove_node(atom, binary) :: {:ok, [{binary, integer}]} | {:error, :node_not_exists}
-  def remove_node(name, node_name) do
-    GenServer.call(name, {:remove_node, node_name})
+  @doc """
+  Adds multiple nodes to the existing set of nodes in the ring.
+  """
+  @spec add_nodes(name :: atom, nodes :: [binary | {binary, integer}]) :: {:ok, [{binary, integer}]} | {:error, :node_exists}
+  def add_nodes(name, nodes) do
+    GenServer.call(name, {:add_nodes, nodes})
   end
 
-  @spec set_overrides(atom, override_map) :: {:ok, override_map}
-  def set_overrides(name, overrides) do
-    GenServer.call(name, {:set_overrides, overrides})
+  @doc """
+  Finds the node responsible for the given key in the specified ring.
+  """
+  @spec find_node(atom, binary | integer) :: {:ok, binary} | {:error, atom}
+  def find_node(name, key) do
+    with {:ok, [node]} <- find_nodes(name, key, 1)  do
+      {:ok, node}
+    end
   end
 
-  @spec get_overrides(atom) :: {:ok, override_map}
-  def get_overrides(name) do
-    GenServer.call(name, :get_overrides)
+  @doc """
+  Finds the specified number of nodes responsible for the given key in the specified ring's current generation.
+  """
+  @spec find_nodes(atom, binary | integer, integer) :: {:ok, [binary]} | {:error, atom}
+  def find_nodes(name, key, num) do
+    with {:ok, {{table, num_nodes}, _previous_table, gen, overrides}} <- Config.get(name) do
+      do_find_nodes_in_table(key, table, overrides, gen, num_nodes, num)
+    end
   end
 
-  @spec get_nodes(atom) :: {:ok, [binary]}
-  def get_nodes(name) do
-    GenServer.call(name, :get_nodes)
+  @doc """
+  Finds the specified number of nodes responsible for the given key in the specified ring's previous generation.
+  """
+  @spec find_previous_nodes(atom, binary | integer, integer) :: {:ok, [binary]} | {:error, atom}
+  def find_previous_nodes(name, key, num) do
+    with {:ok, {_current_table, {table, num_nodes}, gen, overrides}} <- Config.get(name) do
+      do_find_nodes_in_table(key, table, overrides, gen, num_nodes, num)
+    end
   end
 
-  @spec get_nodes_with_replicas(atom) :: {:ok, [{binary, integer}]}
-  def get_nodes_with_replicas(name) do
-    GenServer.call(name, :get_nodes_with_replicas)
+  @doc """
+  Finds the specified number of nodes responsible for the given key in the specified ring's current generation AND in
+  the specified ring's previous generation.  This means that this function returns up to 2 * number of nodes requested.
+  """
+  @spec find_stable_nodes(atom, binary | integer, integer) :: {:ok, [binary]} | {:error, atom}
+  def find_stable_nodes(name, key, num) do
+    with {:ok, current_nodes} <- find_nodes(name, key, num),
+         {:ok, previous_nodes} <- find_previous_nodes(name, key, num) do
+      stable_nodes =
+        previous_nodes
+        |> Enum.reverse()
+        |> Enum.reduce(Enum.reverse(current_nodes), fn node, acc ->
+          if node in acc do
```

Comment:
I'm assuming this doesn't get run often so doing O(n^2) here is okay given that n is < 10?

Reply:
This function will be run a lot, anytime you want to find the stable nodes for a key.  `n` is normally either 2 or 3 in practice because you want to query the primary and secondary (and sometimes tertiary).

I'm open to changing this code, I had considered converting them into MapSets and unioning them together, but that would lose the ordering information which is important here and I suspected that creating two mapsets from the list, doing the union, converting back to a list would end up roughly as expensive for the n=2 or n=3 usecases.

Reply:
Well you only need to have one `MapSet` as part of the acc that you pass (`acc = {list, set}`), but yeah I didn't think it was worth allocating either. Just double checking my assumptions.